_posts/2023-04-30-writing-code-with-chatgpt-improve-it-with-kani.md

```diff
@@ -0,0 +1,369 @@
+We ran Kani on the above program
+
+```console
+> kani integer_average.rs
+```
+
+and got the verification result
+
+```console
+...
+SUMMARY:
+ ** 1 of 4 failed
+Failed Checks: attempt to add with overflow
+ File: "...", line 2, in integer_average
+
+VERIFICATION:- FAILED
+Verification Time: 0.6991158s
+```
+
+The code is actually buggy! 
+Although the average of two `i32` integers is in the range of `i32`, the intermediate sum `a + b` may overflow.
+A counterexample is when `a` equals `i32::MAX` and `b` equals any positive integer, the sum of them will be greater than `i32::MAX`, and hence the overflow.
+
+
+The examples above illustrate that, although code generation is efficient and accessible, it's crucial to exercise caution and avoid blindly trusting the generated code, as it may contain errors.
+We believe there are two factors that contribute to this unreliability.
+Firstly, the training corpus for most LLMs is unverified code.
+Secondly, text prompts can be ambiguous, leading to the production of flawed code.
+Fortunately, program verifiers such as Kani can provide valuable information about generated code, such as verification results, and counterexamples.
+Armed with this information, developers can locate and fix bugs more effectively.
+But what about the potential for language models like ChatGPT to leverage such insights to improve code generation accuracy?
+It's an intriguing question that merits further exploration.
+
+
+
+
+## Code Generation with Verifiers in the Loop
+The last task we explore in this blog is: combining Kani and ChatGPT to generate code that passes verification.
+
+Recall that we found the overflow error in the expression `a + b` in the generated function `integer_average` with Kani.
+We shared this information with ChatGPT.
+
+```
+Q: The expression (a + b) / 2 may overflow.
+```
+
+ChatGPT agreed with us and fixed the overflow with a new implementation using `checked_add`:
+
+``` rust
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    match num1.checked_add(num2) {
+        Some(sum) => sum / 2,
+        None => {
+            // Handle overflow gracefully
+            println!("Overflow occurred!");
+            std::i32::MAX // Fallback value
+        }
+    }
+}
+```
+
+We reran Kani on this new implementation.
+Kani verified that this new version is correct!
+However, the function is no longer complete because it doesn't computes the average for all `i32` integers as we specified.
+For example, it does not compute the average for the input `integer_average(std::i32::MAX, 1)`.
+So, we tried again with the following prompt to generate a complete version of `integer_average`.
+
+```
+Q: Can you give me a fix that does not throw any error?
+```
+
+This time, ChatGPT found us an implementation that accepts all `i32` integers as input.
+
+```
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    num1 / 2 + num2 / 2 + (num1 % 2 + num2 % 2) / 2
+}
+```
+
+Kani verifies that this implementation is correct.
```

Comment:
It only verifies the absence of certain classes of bugs (overflows, panics, etc.). It doesn't really prove that the function is correctly computing the integer average of the two numbers.

Reply:
How about "Kani didn't report any failure on this implementation."